### ISSUE TITLE
Why do I need 'mp.set_start_method('spawn', force=True)' to get ddp to work for multi-GPU training?

### ISSUE THREAD

**Issue by icdh99** [Lightning-AI/lightning#17026]
### Bug description

I recently started using Pytorch Lightning, and want to use multiple GPUs to speed up my model training. Among others I used this example: https://pytorch-lightning.readthedocs.io/en/stable/clouds/cluster_advanced.html#build-your-slurm-script 

Here is a preview of my code:

```
    class model(pl.LightningModule):
	def __init__(self):
		super(model, self).__init__()
		self.linear = nn.Linear(in_features = 3072, out_features = 5313, bias = True)
		self.softplus = nn.Softplus(beta = 1, threshold = 20)	
		self.lr = 1e-4
		self.loss = nn.PoissonNLLLoss()
		self.save_hyperparameters()


	def forward(self, x):
		x = self.linear(x)
		x = self.softplus(x)
		return x

	def configure_optimizers(self): return torch.optim.Adam(self.parameters(), lr = self.lr)

	def training_step(self, train_batch, batch_idx):
		x, y = train_batch 
		logits = self.forward(x)
		loss = self.loss(logits, y)
		self.log("train_loss", loss, on_epoch=True, prog_bar=True, sync_dist=True)
		self.logger.experiment.add_scalars('loss', {'train': loss},self.global_step)
		return loss

	def validation_step(self, valid_batch, batch_idx):
		x, y = valid_batch
		logits = self.forward(x)
		val_loss = self.loss(logits, y)
		self.log("val_loss", val_loss, prog_bar=True, sync_dist=True)
		self.logger.experiment.add_scalars('loss', {'valid': val_loss},self.global_step)
		return val_loss

    class MyDataset(Dataset):
    def __init__(self, list_IDs): self.list_IDs = list_IDs

    def __len__(self): return len(self.list_IDs)
    
    def __getitem__(self, index):
        ID = self.list_IDs[index]
        x = torch.load(f'embeddings_seq{ID}.pt')
        y = torch.load(f'targets_seq{ID}.pt')
        y = torch.squeeze(y) # required for correct shape
        return x, y

# start of file train.py:
    def main():
    mp.set_start_method('spawn', force=True)

    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    print('Using device (set to GPU if available):', device)

    N_PROCESSES = 2
    BATCH_SIZE = 32
    EPOCHS = 1
    strategy = 'ddp'

    training_set = MyDataset(partition_indices['train'])
    training_generator = DataLoader(training_set, batch_size = BATCH_SIZE, shuffle = True, 
                                    num_workers = 2,
                                    pin_memory = False)
    val_set = MyDataset(partition_indices['val'])
    val_generator = DataLoader(val_set, batch_size = BATCH_SIZE, shuffle = False, 
                                    num_workers = 2,
                                    pin_memory = False)

    clf = model()
    trainer = pl.Trainer(max_epochs = EPOCHS, 
                        callbacks = callbacks, 
                        enable_checkpointing = True, 
                        logger = logger, 
                        accelerator = 'gpu', 
                        num_nodes = 1,
                        devices = 2, 
                        precision = 16,
                        strategy=strategy) 
    trainer.fit(clf, training_generator, val_generator)

    if __name__ == "__main__":
    main()

```
part of slurm submit file:
```
#!/bin/bash
#SBATCH --gres=gpu:2
#SBATCH --ntasks-per-node=2
#SBATCH --nodes=1

srun python train.py
```

If I remove the line `mp.set_start_method('spawn', force=True)`, I get the following error:
`RuntimeError: Cannot re-initialize CUDA in forked subprocess. To use CUDA with multiprocessing, you must use the 'spawn' start method`

I would preferably only use strategy = 'ddp' like in https://pytorch-lightning.readthedocs.io/en/stable/accelerators/gpu_intermediate.html, because it states that using spawn is not desirable.

Is this the correct way of implementing the ddp strategy? 

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

* CUDA:
	- GPU:
		- GRID V100S-32Q
	- available:         True
	- version:           11.7
* Lightning:
	- discrete-key-value-bottleneck-pytorch: 0.0.7
	- enformer-pytorch:  0.6.1
	- lightning-utilities: 0.7.0
	- pytorch-lightning: 1.9.3
	- torch:             1.13.1
	- torchaudio:        0.13.1
	- torchmetrics:      0.11.1
	- torchvision:       0.14.1
	- vector-quantize-pytorch: 0.10.15
* Packages:
	- absl-py:           1.3.0
	- aiohttp:           3.8.3
	- aiosignal:         1.2.0
	- appdirs:           1.4.4
	- argcomplete:       2.0.0
	- argh:              0.27.2
	- arrow:             1.2.3
	- astroid:           2.14.2
	- astunparse:        1.6.3
	- async-timeout:     4.0.2
	- attrs:             22.1.0
	- binaryornot:       0.4.4
	- biopython:         1.81
	- blinker:           1.4
	- bottleneck:        1.3.5
	- brotlipy:          0.7.0
	- cachetools:        4.2.2
	- certifi:           2022.12.7
	- cffi:              1.15.1
	- chardet:           5.1.0
	- charset-normalizer: 3.0.1
	- click:             8.0.4
	- colorama:          0.4.6
	- colorlog:          6.7.0
	- contourpy:         1.0.5
	- cookiecutter:      2.1.1
	- cryptography:      38.0.4
	- cycler:            0.11.0
	- deprecation:       2.1.0
	- dill:              0.3.6
	- discrete-key-value-bottleneck-pytorch: 0.0.7
	- einops:            0.6.0
	- enformer-pytorch:  0.6.1
	- filelock:          3.9.0
	- flatbuffers:       2.0
	- flit-core:         3.6.0
	- fonttools:         4.25.0
	- frozenlist:        1.3.3
	- fsspec:            2023.1.0
	- future:            0.18.3
	- gast:              0.4.0
	- gffutils:          0.11.1
	- google-api-core:   2.11.0
	- google-auth:       2.6.0
	- google-auth-oauthlib: 0.4.4
	- google-cloud-core: 2.3.2
	- google-cloud-storage: 2.7.0
	- google-crc32c:     1.5.0
	- google-pasta:      0.2.0
	- google-resumable-media: 2.4.1
	- googleapis-common-protos: 1.58.0
	- grpcio:            1.42.0
	- h5py:              3.8.0
	- huggingface-hub:   0.12.0
	- idna:              3.4
	- isort:             5.12.0
	- jinja2:            3.1.2
	- jinja2-time:       0.2.0
	- joblib:            1.1.1
	- keras:             2.10.0
	- keras-preprocessing: 1.1.2
	- kipoi:             0.8.6
	- kipoi-conda:       0.3.1
	- kipoi-utils:       0.7.7
	- kipoiseq:          0.5.2
	- kiwisolver:        1.4.4
	- lazy-object-proxy: 1.9.0
	- lightning-utilities: 0.7.0
	- markdown:          3.4.1
	- markdown-it-py:    2.2.0
	- markupsafe:        2.1.2
	- matplotlib:        3.6.2
	- mccabe:            0.7.0
	- mdurl:             0.1.2
	- mkl-fft:           1.3.1
	- mkl-random:        1.2.2
	- mkl-service:       2.4.0
	- multidict:         6.0.2
	- munkres:           1.1.4
	- natsort:           7.1.1
	- numexpr:           2.8.4
	- numpy:             1.24.1
	- nvidia-cublas-cu11: 11.10.3.66
	- nvidia-cuda-nvrtc-cu11: 11.7.99
	- nvidia-cuda-runtime-cu11: 11.7.99
	- nvidia-cudnn-cu11: 8.5.0.96
	- oauthlib:          3.2.1
	- opt-einsum:        3.3.0
	- packaging:         23.0
	- pandas:            1.5.2
	- pillow:            9.3.0
	- pip:               22.3.1
	- platformdirs:      3.0.0
	- ply:               3.11
	- polars:            0.16.1
	- pooch:             1.4.0
	- protobuf:          4.21.12
	- pyasn1:            0.4.8
	- pyasn1-modules:    0.2.8
	- pycparser:         2.21
	- pyfaidx:           0.7.1
	- pygments:          2.14.0
	- pyjwt:             2.4.0
	- pylint:            2.16.2
	- pyopenssl:         22.0.0
	- pyparsing:         3.0.9
	- pyqt5-sip:         12.11.0
	- pysocks:           1.7.1
	- python-dateutil:   2.8.2
	- python-slugify:    8.0.0
	- pytorch-lightning: 1.9.3
	- pytz:              2022.7
	- pyyaml:            6.0
	- regex:             2022.10.31
	- related:           0.7.3
	- requests:          2.28.1
	- requests-oauthlib: 1.3.0
	- rich:              13.3.1
	- rsa:               4.7.2
	- scikit-learn:      1.2.1
	- scipy:             1.10.0
	- seaborn:           0.12.2
	- setuptools:        65.6.3
	- simplejson:        3.18.3
	- sip:               6.6.2
	- six:               1.16.0
	- tensorboard:       2.10.0
	- tensorboard-data-server: 0.6.1
	- tensorboard-plugin-wit: 1.8.1
	- tensorboardx:      2.2
	- tensorflow:        2.10.0
	- tensorflow-estimator: 2.10.0
	- termcolor:         2.1.0
	- text-unidecode:    1.3
	- threadpoolctl:     2.2.0
	- tinydb:            4.7.1
	- tokenizers:        0.13.2
	- toml:              0.10.2
	- tomli:             2.0.1
	- tomlkit:           0.11.6
	- torch:             1.13.1
	- torchaudio:        0.13.1
	- torchmetrics:      0.11.1
	- torchvision:       0.14.1
	- tornado:           6.1
	- tqdm:              4.64.1
	- transformers:      4.26.0
	- typing-extensions: 4.4.0
	- urllib3:           1.26.14
	- vector-quantize-pytorch: 0.10.15
	- werkzeug:          2.2.2
	- wheel:             0.37.1
	- wrapt:             1.14.1
	- yarl:              1.8.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.10.9
	- version:           #1 SMP Mon Jan 23 12:23:56 EST 2023

cc @justusschock @awaelchli

**Comment by awaelchli** [comment]
Hey @icdh99 
As best as I can tell, this is because of a combination of three things that don't play well together:
- `num_workers>0` in your DataLoader (which is understandable and normal)
- Your system has the fork start method as default for multiprocessing
- Something, either in your code or in Lightning, is creating a cuda context in the main process before the dataloader forks processes (this is the code we need to find to fix the issue). 

I believe the guilty line of code is this: `device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')`, because this initializes a cuda context. See this issue on pytorch for reference: https://github.com/pytorch/pytorch/issues/83973

I suggest you remove this line of code and instead set `accelerator="auto"` in the Trainer. Let me know how it goes.

**Comment by icdh99** [comment]
Hi,

thank you for your quick response and possible solution! I will try the following out tomorrow and let you know how it goes:

- remove line `device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')`
- keep `num_workers = 2 `
- remove line `mp.set_start_method('spawn', force=True)`

The tensors that I load in are stored on the gpu by default. I will see if that throws an error and if so, I will change it into this:
- `torch.load('tensors.pt', map_location=torch.device('cpu'))`
- this means I might be able to use the` pin_memory = True` option as well

**Comment by awaelchli** [comment]
Yes, good idea. Please let me know if that helped :)

**Comment by icdh99** [comment]
Hi!

I implemented the changes I described above (including setting `accelator = "auto"` which I forgot to mention in my previous comment). I had to load my tensors to the cpu like I described. 

The model runs now without the `mp.set_start_method('spawn', force=True)`, so thank you for the fix! I did not think about the cuda context so thank you for pointing me to that :). 

Do you have a recommendation for the value of the num_workers argument? I am currently using two gpu's with this request:
```
#SBATCH --gres=gpu:2
#SBATCH --ntasks-per-node=2
#SBATCH --nodes=1
```

I am matching this with the following arguments in the Trainer:
--nodes = 1 --> Trainer(num_nodes = 1)
--ntasks-per-node=2 --> Trainer(devices = 2)

I am limited to requesting this with our cluster, so if you might have any other ideas how to maximise the use of these gpu's I would be happy to hear!

**Comment by awaelchli** [comment]
If you think your training speed is bottlenecked by the data loading, then increasing num_workers can help sometimes. I recommend setting it to `num_workers = #cpu cores / #gpus`. There are many other things that can help speed up your training, for example: using mixed precision (precision="bf16"), increasing batch size to maximize throughput, using PyTorch 2.0 and torch.compile.

**Comment by icdh99** [comment]
thank you, will look into those!

**Comment by icdh99** [comment]
Hi! I have implemented the changes as described above. After training the complete model again I noticed that the complete training time has increased from 3 days to 4 days for the same number of epochs etc. Is that expected behaviour with the implemented changes? 

The num_workers is kept at 2 as I cannot increase the number due to memory constraints. 

The results are similar and I did not change anything for the input data/model architecture/anything really!
